File: demo/ebpf_trace/trace/trace_fmt.py
from bcc import BPF
from time import strftime

ELF_FILE_PATH = b"/home/gan/code/go/my_code/go-exp/demo/ebpf_trace/ebpf_trace"

b = BPF(src_file=b"trace_fmt.c")

b.attach_uprobe(name=ELF_FILE_PATH,
                sym=b"os.open", fn_name=b"go_open_file")

b.attach_uprobe(name=ELF_FILE_PATH,
                sym=b"os.(*File).Write", fn_name=b"go_write_data")


def print_event(cpu, data, size):
    event = b["events"].event(data)
    print("%-9s %-6d arg1:%-20s, arg2:%-8d, arg3:%-4d" %
          (strftime("%H:%M:%S"), event.uid, event.arg1[:event.ln], event.arg2, event.arg3))


print("%-9s %-6s %-26s %-16s %-16s" % ("TIME", "PID", "ARG1", "ARG2", "ARG3"))

b["events"].open_perf_buffer(print_event)

while True:
    try:
        b.perf_buffer_poll()
    except KeyboardInterrupt:
        exit()

File: demo/ebpf_trace/trace/trace_fmt.c
#include <uapi/linux/ptrace.h>

struct data_t
{
    u32 uid;
    char arg1[20];
    int ln;
    int arg2;
    int arg3;
};

BPF_PERF_OUTPUT(events);

int go_open_file(struct pt_regs *ctx)
{
    // arg1: {RAX, RBX}, string
    // arg2: RCX, int
    // arg3: RDX, int
    struct data_t data = {};
    data.uid = bpf_get_current_uid_gid();

    // 读取用户空间的字符串到data
    u32 size = (u32)(ctx->bx);
    bpf_probe_read_user(&data.arg1, sizeof(data.arg1), (void *)ctx->ax);
    data.ln = size;
    data.arg2 = ctx->cx;
    data.arg3 = ctx->di;
    events.perf_submit(ctx, &data, sizeof(data));
    return 0;
}

int go_write_data(struct pt_regs *ctx)
{
    // receiver: RAX
    // arg1: {RBX, RCX, RDI}, []byte
    struct data_t data = {};
    data.uid = bpf_get_current_uid_gid();

    // 读取用户空间的字符串到data
    u32 size = (u32)(ctx->cx);
    bpf_probe_read_user(&data.arg1, sizeof(data.arg1), (void *)ctx->bx);
    data.ln = size;
    events.perf_submit(ctx, &data, sizeof(data));
    return 0;
}